### ISSUE TITLE
Investigate Kotlin 1.6

### ISSUE THREAD

**Issue by martinbonnin** [apollographql/apollo-android#3555]
Kotlin 1.6 is out 🎉 . Investigate what's interesting and if we should switch. One question is whether users on 1.4.32 will be able to compile against `apollo-android` compiled with 1.6

**Comment by joreilly** [comment]
Possibly a lot more work but just in case you want to take approach that https://github.com/rickclephas/KMP-NativeCoroutines did and have different version with particular suffix that supports 1.6. btw I did quick test here at updating projects I have here that are using apollo to use 1.6 and getting some crashes though they could easily be related to something else....need to dig deeper (not sure if it would be expected that current version would work with 1.6. or not?).

**Comment by joreilly** [comment]
I wonder if new kotlin/native memory model option would also need to be enabled by each library for overall project to use it...

**Comment by martinbonnin** [comment]
> have different version with particular suffix that supports 1.6

Ohoo interesting 👀 . I don't want to think of the astronomical number of artifacts this will upload to mavenCentral but that's super clever 👍 

> not sure if it would be expected that current version would work with 1.6. or not?

My naive expectation is that Kotlin 1.6 should be backward compatible. So a project compiled with 1.6 "should" work with "1.5" libs. Of course native is not as stable so not sure. If you have a branch somewhere, feel free to share it, I'll look into it.

> I wonder if new kotlin/native memory model option would also need to be enabled by each library for overall project to use it

Interesting question... I was under the impression that this was a runtime-only thing but haven't really digged into it much.

**Comment by joreilly** [comment]
Pushed changes I've made so far to following branch of StarWars sample - https://github.com/joreilly/StarWars/tree/kotlin_1_6.  I started with that one as it has least  number of dependencies

Using following approach of using Kotlin 1.6.0 when building iOS app (as Jetpack Compose still depends on 1.5.31)....iOS app builds fine but crashes on startup.

```
./gradlew :shared:embedAndSignAppleFrameworkForXcode -P"kotlinVersion"="1.6.0"
```

**Comment by joreilly** [comment]
Hmm, seems to run fine on first run after clean build in XCode but crashes on subsequent runs.....

**Comment by joreilly** [comment]
ok, looks like issue I was having was related to incorrect version of kotlinx coroutines I was using.

**Comment by martinbonnin** [comment]
After reading https://youtrack.jetbrains.com/issue/KT-42293, I think as a library we shouldn't rush too much into 1.6 as it will force Kotlin Native consumers to use Kotlin 1.6 before updating to the latest version of Apollo Android (native has no forward compatibility. JVM has one version of forward compatibility). 

Updating is always a good thing but leaving a couple of weeks of head start sounds better for everyone here?

**Comment by BoD** [comment]
> After reading https://youtrack.jetbrains.com/issue/KT-42293, I think as a library we shouldn't rush too much into 1.6 as it will force Kotlin Native consumers to use Kotlin 1.6 before updating to the latest version of Apollo Android (native has no forward compatibility. JVM has one version of forward compatibility).
> 
> Updating is always a good thing but leaving a couple of weeks of head start sounds better for everyone here?

Seems reasonable to me 👍

**Comment by martinbonnin** [comment]
Closed with #3563
